### PR TITLE
[MM-55605] Modified server/../app/bot.go public methods with request context

### PR DESCRIPTION
**Pull request by ruvyas**
#### Summary

- Add request context to all public methods with *App receiver in bot.go
- Removed GetWarnMetricsBot as it is unused
- Resolves https://github.com/mattermost/mattermost/issues/25494
- Closed previous PR and re-opened (https://github.com/mattermost/mattermost/pull/25615)

#### Ticket Link

Jira https://mattermost.atlassian.net/browse/MM-55605

#### Screenshots

Not Applicable

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by ruvyas** [comment]
@hanzei I did notice other files which references `GetBot(s)` and couple other methods. I wasn't sure how to propagate `context` there.

**Comment by fmartingr** [comment]
> @hanzei I did notice other files which references `GetBot(s)` and couple other methods. I wasn't sure how to propagate `context` there.

One solution would be to also add `request.CTX` in those methods as well, from the first error I see in CI:

```go
func (a *App) SessionHasPermissionToManageBot(session model.Session, botUserId string) *model.AppError {
	existingBot, err := a.GetBot(botUserId, true)
```

would become:

```go
func (a *App) SessionHasPermissionToManageBot(rctx request.CTX, session model.Session, botUserId string) *model.AppError {
	existingBot, err := a.GetBot(rctx, botUserId, true)
```

One place this is being called is `channels/api4/bot.go`:

```go
if err := c.App.SessionHasPermissionToManageBot(*c.AppContext.Session(), botUserId); err != nil {
```

This would become:

```go
if err := c.App.SessionHasPermissionToManageBot(c.AppContext, *c.AppContext.Session(), botUserId); err != nil {
```

You would probably need to go up in the chain more (where other functions are being called) and we may follow the same approach on those as well if we want to properly propagate the context.

is this approach good @hanzei ?

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking until the build issues have been fixed

File: server/channels/app/server.go [hunk 1]
```diff
@@ -1825,14 +1825,14 @@ func runDNDStatusExpireJob(a *App) {
 func runPostReminderJob(a *App) {
 	if a.IsLeader() {
 		withMut(&a.ch.postReminderMut, func() {
-			a.ch.postReminderTask = model.CreateRecurringTaskFromNextIntervalTime("Check Post reminders", a.CheckPostReminders, 5*time.Minute)
+			a.ch.postReminderTask = model.CreateRecurringTaskFromNextIntervalTime("Check Post reminders", a.CheckPostReminders(request.EmptyContext(a.Log())), 5*time.Minute)
```
Comment (hanzei) [inline server/channels/app/server.go:1828]:
```suggestion
			a.ch.postReminderTask = model.CreateRecurringTaskFromNextIntervalTime("Check Post reminders", a.CheckPostReminders(request.EmptyContext(a.Log()), 5*time.Minute)
```

File: server/channels/app/server.go [hunk 2]
```diff
@@ -1825,14 +1825,14 @@ func runDNDStatusExpireJob(a *App) {
...
 		})
 	}
 	a.ch.srv.AddClusterLeaderChangedListener(func() {
 		mlog.Info("Cluster leader changed. Determining if post reminder task should be running", mlog.Bool("isLeader", a.IsLeader()))
 		if a.IsLeader() {
 			withMut(&a.ch.postReminderMut, func() {
-				a.ch.postReminderTask = model.CreateRecurringTaskFromNextIntervalTime("Check Post reminders", a.CheckPostReminders, 5*time.Minute)
+				a.ch.postReminderTask = model.CreateRecurringTaskFromNextIntervalTime("Check Post reminders", a.CheckPostReminders(request.EmptyContext(a.Log())), 5*time.Minute)
```
Comment (hanzei) [inline server/channels/app/server.go:1835]:
```suggestion
				a.ch.postReminderTask = model.CreateRecurringTaskFromNextIntervalTime("Check Post reminders", a.CheckPostReminders(request.EmptyContext(a.Log()), 5*time.Minute)
```

File: server/channels/app/server.go [hunk 3]
```diff
@@ -1825,14 +1825,14 @@ func runDNDStatusExpireJob(a *App) {
 func runPostReminderJob(a *App) {
 	if a.IsLeader() {
 		withMut(&a.ch.postReminderMut, func() {
```
Comment (hanzei) [inline server/channels/app/server.go:1827]: Can all methods in `runPostReminderJob` share a `request.CTX`?

Comment (ruvyas) [reply server/channels/app/server.go:1827]: I assume `request.CTX` will be different if cluster leader changes?

Comment (hanzei) [reply server/channels/app/server.go:1827]: No, it shouldn't be different.

**Comment by hanzei** [comment]
@ruvyas Please let me know if you need any further help fixing the remaining build issues.

**Comment by ruvyas** [comment]
@hanzei I think I will need help with this PR. Looks like `CheckPostReminders` in `server.go` syntax isn't correct.

**Comment by hanzei** [comment]
@ruvyas I've fixed the syntax and merged `master` into your branch. Let's see what CI has to say.

**Pull request closed without merging**